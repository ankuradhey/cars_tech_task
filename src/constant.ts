export const constant = {
    host: "https://auto1-mock-server.herokuapp.com",
    carsUrl: "/api/cars",
    colorsUrl: "/api/colors",
    manufacturersUrl: "/api/manufacturers",
    carDetailsUrl: "/api/cars/{stockNumber}",
};
