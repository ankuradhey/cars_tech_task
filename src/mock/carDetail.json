{
    "car": {
        "stockNumber": 10034,
        "manufacturerName": "Fiat",
        "modelName": "Grande Punto",
        "color": "black",
        "mileage": {
            "number": 170648,
            "unit": "km"
        },
        "fuelType": "Diesel",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"
    }
}