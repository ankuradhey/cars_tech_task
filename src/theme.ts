import { DefaultTheme } from "styled-components";

export const defaultTheme: DefaultTheme = {
    borderRadius: "2px",
    palette: {
        common: {
            black: "#4A4A4A",
            white: "#EDEDED",
        },
        primary: {
            main: "#EA7F28",
            contrastText: "#ffffff",
        },
        secondary: {
            main: "#D37324",
            contrastText: "#ffffff",
        },
        tertiary: {
            main: "#ffffff",
            contrastText: "#EDEDED",
        },
    },
    gutterSize: {
        sm: "8px",
        md: "12px",
        lg: "24px",
    },
};
