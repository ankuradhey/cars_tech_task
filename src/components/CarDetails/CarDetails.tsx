import React, { FC, useState, useEffect } from "react";
import { Card, Col, Container, Row } from "react-bootstrap";
import { useParams } from "react-router-dom";
import { constant } from "../../constant";
import { ImgCard } from "../ImgCard";
import { Text } from "../Text";
import { Car } from "../../types";
import { Button } from "../Button";

export const CarDetails: FC = () => {
    let { id } = useParams();
    const savedCars: number[] = JSON.parse(localStorage.getItem("savedCars") || "[]");
    const [saved, setSaved] = useState(!!savedCars.find((val) => val === id));
    const [carData, setCarData] = useState<Car>(null);

    const saveCar = () => {
        let newSavedCars = [...savedCars, id];
        localStorage.setItem("savedCars", JSON.stringify(newSavedCars));
        setSaved(true);
    };

    const unSaveCar = () => {
        let newSavedCars = savedCars.filter((val) => val !== id);
        localStorage.setItem("savedCars", JSON.stringify(newSavedCars));
        setSaved(false);
    };

    const saveClick = () => {
        if (saved) {
            unSaveCar();
        } else {
            saveCar();
        }
    };

    useEffect(() => {
        const fetchData = async () => {
            const result = await fetch(
                `${constant.host}${constant.carDetailsUrl.replace("{stockNumber}", id)}`
            );
            const data = await result.json();
            setCarData(data.car);
        };
        fetchData();
    }, [id]);

    if (!carData) return <Text size="large">Loading...</Text>;

    const {
        manufacturerName,
        modelName,
        pictureUrl,
        stockNumber,
        mileage,
        fuelType,
        color,
    } = carData;

    return (
        <>
            <Container fluid className="margin-bottom-3">
                <Row>
                    <Col>
                        <ImgCard src={pictureUrl} width="100%" height="400px" />
                    </Col>
                </Row>
            </Container>
            <Container style={{ width: "800px" }}>
                <Row>
                    <Col md={7}>
                        <Text size="heading" className="d-block margin-bottom-3">
                            {manufacturerName} {modelName}
                        </Text>
                        <Text size="large" className="d-block margin-bottom-3">
                            Stock # {stockNumber} - {mileage?.number} {mileage?.unit?.toUpperCase()}{" "}
                            - {fuelType} - {color}
                        </Text>
                        <Text size="regular" className="d-block margin-bottom-3 text-justify">
                            This car is currently available and can be delivered as sson as tomorrow
                            morning. Please be aware that delivery times shown in this page are not
                            definitive and may change due to bad weather conditions.
                        </Text>
                    </Col>
                    <Col md={5}>
                        <Card>
                            <Card.Body>
                                <Text
                                    size="regular"
                                    className="d-block margin-bottom-1 text-justify"
                                >
                                    If you like this car, click the button and save it in your
                                    collection of favorite items.
                                </Text>
                                {saved ? (
                                    <Button
                                        variant="tertiary"
                                        className="float-right"
                                        onClick={saveClick}
                                    >
                                        Unsave
                                    </Button>
                                ) : (
                                    <Button
                                        variant="primary"
                                        className="float-right"
                                        onClick={saveClick}
                                    >
                                        Save
                                    </Button>
                                )}
                            </Card.Body>
                        </Card>
                    </Col>
                </Row>
            </Container>
        </>
    );
};
