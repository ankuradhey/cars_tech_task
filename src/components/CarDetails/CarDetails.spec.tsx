import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { Router, Route } from "react-router-dom";
import { createMemoryHistory } from "history";
import nock from "nock";
import { CarDetails } from "./CarDetails";
import { ThemeProvider } from "styled-components";
import { defaultTheme } from "../../theme";
import * as carDetailMock from "../../mock/carDetail.json";
import { constant } from "../../constant";

describe("CarDetails Component", () => {
    beforeEach(() => {
        nock(constant.host)
            .defaultReplyHeaders({
                "access-control-allow-origin": "*",
                "access-control-allow-credentials": "true",
            })
            .get(`${constant.carDetailsUrl.replace("{stockNumber}", "10058")}`)
            .reply(200, carDetailMock);
    });

    it("Should render CarDetails component", async () => {
        const history = createMemoryHistory();
        const route = "/car/10058";
        history.push(route);

        const { container, getByText } = render(
            <ThemeProvider theme={defaultTheme}>
                <Router history={history} initialEntries={["/car/10058"]}>
                    <Route path="/car/:id" component={CarDetails} />
                </Router>
            </ThemeProvider>
        );
        expect(getByText("Loading...")).toBeInTheDocument();
        await waitFor(() =>
            screen.findByText(
                `${carDetailMock.car.manufacturerName} ${carDetailMock.car.modelName}`
            )
        );
        expect(
            getByText(`${carDetailMock.car.manufacturerName} ${carDetailMock.car.modelName}`)
        ).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
