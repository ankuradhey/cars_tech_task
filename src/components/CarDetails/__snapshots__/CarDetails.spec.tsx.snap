// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarDetails Component Should render CarDetails component 1`] = `
<div>
  <div
    class="margin-bottom-3 container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="sc-bdfBwQ dWa-Dpq"
          height="400px"
          loading="false"
          width="100%"
        >
          <img
            alt="car"
            src="https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="container"
    style="width: 800px;"
  >
    <div
      class="row"
    >
      <div
        class="col-md-7"
      >
        <div
          class="sc-gsTCUz fHgoa-D d-block margin-bottom-3"
        >
          Fiat
           
          Grande Punto
        </div>
        <div
          class="sc-gsTCUz kEImGf d-block margin-bottom-3"
        >
          Stock # 
          10034
           - 
          170648
           
          KM
           
          - 
          Diesel
           - 
          black
        </div>
        <div
          class="sc-gsTCUz iUuijv d-block margin-bottom-3 text-justify"
        >
          This car is currently available and can be delivered as sson as tomorrow morning. Please be aware that delivery times shown in this page are not definitive and may change due to bad weather conditions.
        </div>
      </div>
      <div
        class="col-md-5"
      >
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <div
              class="sc-gsTCUz iUuijv d-block margin-bottom-1 text-justify"
            >
              If you like this car, click the button and save it in your collection of favorite items.
            </div>
            <button
              class="sc-dlfnbm bqlYPG float-right"
            >
              Save
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
